mdm/enroll/service.go

```diff
@@ -164,148 +179,149 @@ func (svc *service) Enroll(ctx context.Context) (profile.Mobileconfig, error) {
+	mdmPayload.CheckInURL = svc.URL + mdmPayloadCheckInEndpoint
+	mdmPayload.CheckOutWhenRemoved = true
+	mdmPayload.AccessRights = 8191
 
 	svc.mu.Lock()
-	topic := svc.Topic
+	mdmPayload.Topic = svc.Topic
 	svc.mu.Unlock()
 
-	mdmPayloadContent := MDMPayloadContent{
-		Payload:             *mdmPayload,
-		AccessRights:        allRights(),
-		CheckInURL:          svc.URL + "/mdm/checkin",
-		CheckOutWhenRemoved: true,
-		ServerURL:           svc.URL + "/mdm/connect",
-		Topic:               topic,
-		SignMessage:         true,
-		ServerCapabilities:  []string{perUserConnections, bootstrapToken},
-	}
-
-	payloadContent := []interface{}{}
+	mdmPayload.SignMessage = true
+	mdmPayload.ServerCapabilities = []string{perUserConnections, bootstrapToken}
 
 	if svc.SCEPURL != "" {
-		scepContent := SCEPPayloadContent{
+		scepPayload := cfgprofiles.NewSCEPPayload(EnrollmentProfileId + ".scep")
+		scepPayload.PayloadDescription = scepPayloadDescription
+		scepPayload.PayloadDisplayName = scepPayloadDisplayName
+		scepPayload.PayloadOrganization = profilePayloadOrganization
+
+		scepPayload.PayloadContent = cfgprofiles.SCEPPayloadContent{
 			URL:      svc.SCEPURL,
-			Keysize:  2048,
+			KeySize:  2048,
 			KeyType:  "RSA",
 			KeyUsage: int(x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment),
 			Name:     "Device Management Identity Certificate",
 			Subject:  svc.SCEPSubject,
 		}
 
-		if svc.SCEPChallengeStore != nil {
-			challenge, err := svc.SCEPChallengeStore.SCEPChallenge()
-			if err != nil {
-				return *profile, err
-			}
-			scepContent.Challenge = challenge
-		} else if svc.SCEPChallenge != "" {
-			scepContent.Challenge = svc.SCEPChallenge
+		var err error
+		scepPayload.PayloadContent.Challenge, err = svc.scepChallenge()
+		if err != nil {
+			return nil, err
 		}
 
-		scepPayload := NewPayload("com.apple.security.scep")
-		scepPayload.PayloadDescription = "Configures SCEP"
-		scepPayload.PayloadDisplayName = "SCEP"
-		scepPayload.PayloadIdentifier = EnrollmentProfileId + ".scep"
-		scepPayload.PayloadOrganization = "MicroMDM"
-		scepPayload.PayloadContent = scepContent
-		scepPayload.PayloadScope = "System"
-
-		payloadContent = append(payloadContent, *scepPayload)
-		mdmPayloadContent.IdentityCertificateUUID = scepPayload.PayloadUUID
+		profile.AddPayload(scepPayload)
+		mdmPayload.IdentityCertificateUUID = scepPayload.PayloadUUID
 	}
 
-	payloadContent = append(payloadContent, mdmPayloadContent)
+	profile.AddPayload(mdmPayload)
 
 	// Client needs to trust us at this point if we are using a self signed certificate.
 	if len(svc.TLSCert) > 0 {
-		tlsPayload := NewPayload("com.apple.security.pem")
+		tlsPayload := cfgprofiles.NewCertificatePKCS1Payload(EnrollmentProfileId + ".cert.selfsigned")
 		tlsPayload.PayloadDisplayName = "Self-signed TLS certificate for MicroMDM"
 		tlsPayload.PayloadDescription = "Installs the TLS certificate for MicroMDM"
 		tlsPayload.PayloadIdentifier = EnrollmentProfileId + ".cert.selfsigned"
```

Comment:
This should be redundant now because the ID is set above.

Reply:
Er.. sorry I'm an idiot... which ID is set where? ;) This is for the separate included TLS profile payload in the profile.

Reply:
Oh sorry I see it now I think. :) the duplicated identifier.

Reply:
I thought it would highlight that particular line :)

Reply:
But, yeah. The `cfgprofiles.NewCertificatePKCS1Payload` sets the ID, then `tlsPayload.PayloadIdentifier = EnrollmentProfileId + ".cert.selfsigned"` sets it again.